### PR TITLE
Add applied result in aggregated status

### PR DESCRIPTION
**Pull request by pigletfly**
Signed-off-by: pigletfly <wangbing.adam@gmail.com>

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes https://github.com/karmada-io/karmada/issues/575

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/util/helper/workstatus.go [hunk 1]
```diff
@@ -84,22 +85,34 @@ func assembleWorkStatus(c client.Client, selector labels.Selector, workload *uns
...
+		clusterName, err := names.GetClusterName(work.Namespace)
+		if err != nil {
+			klog.Errorf("Failed to get clusterName from work namespace %s. Error: %v.", work.Namespace, err)
+			return nil, err
+		}
+
+		// if sync work to member cluster failed, then set status back to resource binding.
+		for _, c := range work.Status.Conditions {
+			if c.Status == v1.ConditionFalse {
+				aggregatedStatus := workv1alpha1.AggregatedStatusItem{
```
Comment (mrlihanbo) [inline pkg/util/helper/workstatus.go:97]: It is better to to check the type of condition although we just use 'WorkApplied' type now.

File: pkg/util/helper/workstatus.go [hunk 2]
```diff
@@ -84,22 +85,34 @@ func assembleWorkStatus(c client.Client, selector labels.Selector, workload *uns
...
 			return nil, err
 		}
+		clusterName, err := names.GetClusterName(work.Namespace)
+		if err != nil {
+			klog.Errorf("Failed to get clusterName from work namespace %s. Error: %v.", work.Namespace, err)
+			return nil, err
+		}
+
+		// if sync work to member cluster failed, then set status back to resource binding.
+		for _, c := range work.Status.Conditions {
```
Comment (RainbowMango) [inline pkg/util/helper/workstatus.go:95]: The loop variable `c` is duplicated with parameter `c client.Client`.

File: pkg/util/helper/workstatus.go [hunk 3]
```diff
@@ -84,22 +85,34 @@ func assembleWorkStatus(c client.Client, selector labels.Selector, workload *uns
...
 				return nil, err
 			}
 			if equal {
-				clusterName, err := names.GetClusterName(work.Namespace)
-				if err != nil {
-					klog.Errorf("Failed to get clusterName from work namespace %s. Error: %v.", work.Namespace, err)
-					return nil, err
-				}
-
 				aggregatedStatus := workv1alpha1.AggregatedStatusItem{
```
Comment (mrlihanbo) [inline pkg/util/helper/workstatus.go:112]: ManifestStatuses will be empty for resource without status. For these resource, we should collect apply result

File: pkg/util/helper/workstatus.go [hunk 4]
```diff
@@ -84,22 +85,34 @@ func assembleWorkStatus(c client.Client, selector labels.Selector, workload *uns
...
+
+		// if sync work to member cluster failed, then set status back to resource binding.
+		for _, c := range work.Status.Conditions {
+			if c.Status == v1.ConditionFalse {
+				aggregatedStatus := workv1alpha1.AggregatedStatusItem{
+					ClusterName:    clusterName,
+					Applied:        false,
+					AppliedMessage: c.Message,
+				}
+				statuses = append(statuses, aggregatedStatus)
```
Comment (RainbowMango) [inline pkg/util/helper/workstatus.go:102]: This will append an `extra items, right?

Comment (RainbowMango) [reply pkg/util/helper/workstatus.go:102]: Just draft one
```diff
diff --git a/pkg/util/helper/workstatus.go b/pkg/util/helper/workstatus.go
index dfdd746d..74c1a298 100644
--- a/pkg/util/helper/workstatus.go
+++ b/pkg/util/helper/workstatus.go
@@ -6,7 +6,7 @@ import (
        "reflect"

        "k8s.io/apimachinery/pkg/api/meta"
-       v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+       metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
        "k8s.io/apimachinery/pkg/labels"
        "k8s.io/apimachinery/pkg/runtime/schema"
@@ -103,6 +103,22 @@ func assembleWorkStatus(c client.Client, selector labels.Selector, workload *uns
                        }
                }

+               var applied bool
+               var appliedMsg string
+               if cond := meta.FindStatusCondition(work.Status.Conditions, workv1alpha1.WorkApplied); cond != nil {
+                       switch cond.Status {
+                       case metav1.ConditionTrue:
+                               applied = true
+                       case metav1.ConditionUnknown:
+                               fallthrough
+                       case metav1.ConditionFalse:
+                               applied = false
+                               appliedMsg = cond.Message
+                       default: // should not happen unless the condition api changed.
+                               panic("unexpected status")
+                       }
+               }
+
                for _, manifestStatus := range work.Status.ManifestStatuses {
                        equal, err := equalIdentifier(&manifestStatus.Identifier, identifierIndex, workload)
                        if err != nil {
@@ -110,9 +126,10 @@ func assembleWorkStatus(c client.Client, selector labels.Selector, workload *uns
                        }
                        if equal {
                                aggregatedStatus := workv1alpha1.AggregatedStatusItem{
-                                       ClusterName: clusterName,
-                                       Status:      manifestStatus.Status,
-                                       Applied:     true,
+                                       ClusterName:    clusterName,
+                                       Status:         manifestStatus.Status,
+                                       Applied:        applied,
+                                       AppliedMessage: appliedMsg,
                                }
                                statuses = append(statuses, aggregatedStatus)
                                break
```

File: pkg/util/helper/workstatus.go [hunk 5]
```diff
@@ -6,6 +6,7 @@ import (
 	"reflect"
 
 	"k8s.io/apimachinery/pkg/api/meta"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (RainbowMango) [inline pkg/util/helper/workstatus.go:9]:
```suggestion
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```

**Review by RainbowMango — COMMENTED** [review]
cc @kevin-wangzefeng @irfanurrehman 
Would like to take a look at the API change?

File: pkg/util/helper/workstatus.go (hunk 3 above)
Comment (RainbowMango) [reply pkg/util/helper/workstatus.go:112]: Wow, is that true?

**Comment by RainbowMango** [comment]
Just a reminder.
We are going to cut a new release early next week, I really want this to be included.

File: pkg/util/helper/workstatus.go (hunk 4 above)
Comment (pigletfly) [reply pkg/util/helper/workstatus.go:102]: actually, if failed to sync workload to member cluster, there would be no `ManifestStatuses` in work object.

**Comment by RainbowMango** [comment]
/approve
leave LGTM to @mrlihanbo

**Pull request merged**